### PR TITLE
TESTBED: Avoid extra delays when video decoding is slow

### PR DESCRIPTION
**Pull request by ccawley2011**

### CODE REVIEW

File: video/video_decoder.cpp [hunk 1]
```diff
@@ -112,6 +112,15 @@ bool VideoDecoder::needsUpdate() const {
 	return false;
 }
 
+void VideoDecoder::delayMillis(uint msecs) {
+	if (!needsUpdate())
+		g_system->delayMillis(MIN<uint>(msecs, getTimeToNextFrame()));
+#ifdef __DS__
```
Comment (mikrosk) [inline video/video_decoder.cpp:118]: `ATARI` would benefit from this as well, I'm thinking whether this shouldn't be specified in a more generic way.

File: video/video_decoder.cpp [hunk 2]
```diff
@@ -112,6 +112,15 @@ bool VideoDecoder::needsUpdate() const {
...
+	else
+		g_system->delayMillis(0); /* This is needed to keep the mixer and timers active */
```
Comment (mikrosk) [inline video/video_decoder.cpp:120]: Wouldn't be safer to put `1` here? IIRC some backends (`ATARI` included) use something like
```C
const uint32 threshold = getMillis() + msecs;
	while (getMillis() < threshold) {
		update();
	}
```
so `0` here wouldn't execute it in every case. Perhaps a better solution is possible here as well (like an addition to OSystem API to do exactly that, update the mixer and timers).

**Comment by bluegr** [comment]
@ccawley2011 So, based on @mikrosk's comments, would it make sense to adapt `VideoDecoder::delayMillis()` like this?

    void VideoDecoder::delayMillis(uint msecs) {
        if (!needsUpdate())
            g_system->delayMillis(MIN<uint>(msecs, getTimeToNextFrame()));
        else
            g_system->delayMillis(1);
    }

File: video/video_decoder.cpp [hunk 3]
```diff
@@ -112,6 +112,15 @@ bool VideoDecoder::needsUpdate() const {
...
+#endif
```
Comment (bluegr) [inline video/video_decoder.cpp:121]:
```suggestion
	else
		g_system->delayMillis(1); /* This is needed to keep the mixer and timers active */
```

**Comment by bluegr** [comment]
Adapted @ccawley2011's code as per @mikrosk's review

**Comment by ccawley2011** [comment]
The idea of the PR is that it shouldn't have to delay at all if the frame is taking too long to decode. The reason for calling `delayMillis` with a value of 0 is so that the timer and mixer can be updated regularly (since threads aren't an option), so always having a 1ms delay seems less than ideal.

File: video/video_decoder.cpp (hunk 2 above)
Comment (lephilousophe) [reply video/video_decoder.cpp]: I propose to use this instead:
```
const uint32 threshold = getMillis() + msecs;
do {
    update();
} while (getMillis() < threshold);
```
This makes sure the update is done once when calling delayMillis.

**Comment by lephilousophe** [comment]
I looked at our implementations:
- Atari (as described by @mikrosk) checks timeout before updating timers and mixers,
- Dreamcast does the same,
- DS updates timers and mixers once at start but then busy loops,
- libretro does complex stuff but checks timeout before updating timers (no mixer update),
- N64 updates audio but not timers (checked when polling events),
- other platforms rely on `usleep` or equivalent and OS multithreading.

I think all cases (except the last one) may have to be modified to update audio and timers during the wait with at least one update.

**Comment by sev-** [comment]
Thanks!

**Pull request merged**